packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Providers/CanonicalTypeProvider.cs

```diff
@@ -111,31 +114,144 @@ protected override PropertyProvider[] BuildProperties()
                         customProperty.Type.IsPublic,
                         customProperty.Type.IsStruct,
                         customProperty.Type.BaseType,
-                        TypeFactory.CreatePrimitiveCSharpTypeCore(inputEnumType.ValueType));
+                        TypeFactory.CreatePrimitiveCSharpTypeCore(inputEnumType!.ValueType));
                 }
             }
 
             return [..generatedProperties, ..customProperties];
         }
 
-        private static Dictionary<InputModelProperty, PropertyProvider> BuildSpecToCustomPropertyMap(
-            IReadOnlyList<PropertyProvider> customProperties,
-            Dictionary<string, InputModelProperty> specPropertiesMap)
+        protected override FieldProvider[] BuildFields()
+        {
+            var generatedFields = _generatedTypeProvider.Fields;
+            var customFields = _generatedTypeProvider.CustomCodeView?.Fields ?? [];
+
+            // Update the serializedName of generated properties if necessary
+            foreach (var generatedField in generatedFields)
+            {
+                if (_serializedNameMap.TryGetValue(generatedField.Name, out var serializedName) && serializedName != null)
+                {
+                    generatedField.WireInfo!.SerializedName = serializedName;
+                }
+            }
+
+            foreach (var customField in customFields)
```

Comment:
This looks similar to the BuildProperties method. Do you think it's worth doing some consolidation between the two methods or are the types different enough to where it's worth having this logic live in each method?

Reply:
I considered adding an internal interface to make the code more reusable but we wouldn't be able to use this in the SCM plugin. Since we are mutating the instances it is hard to consolidate the code without them sharing a common base type/interface.